### ISSUE TITLE
upsert across HTTP requests has a race condition

### ISSUE THREAD

**Issue by henrikuper** [prisma/prisma#3242]
## Bug description
Situation: I have a model with a unique string. 
At the beginning I have an empty database. I want to create or update an object of the model with upsert and check if the object exists by searching for the unique string. 

If I call several upserts with the same unique string at the same time, I expect it to be created once and then updated if necessary. But it is created once and then the error occurs: Unique constraint failed on the fields.

```
Unique constraint failed on the fields: (`testId`).
```

## Expected behavior
 I expect it to be created once and then updated the just created object, instead of crashing.
## Prisma information
```
model Test {
  id Int      @default(autoincrement()) @id
  testId     String   @unique
}
const prismaClient = new PrismaClient({
  log: ["info", "warn", "error"],
});

const upsertPromises: Promise<Test>[] = [];
for (let i = 0; i < 2; i++) {
  upsertPromises.push(
    prismaClient.test.upsert({
      where: { testId: "testId" },
      create: {
        testId: "testId",
      },
      update: {},
    })
  );
}
await Promise.all(upsertPromises);

```

## Environment & setup
- OS: Mac OS, 
- Database: PostgreSQL
- Node.js version: v13.3.0
- Prisma version: 2.3.0

```

```

**Comment by pantharshit00** [comment]
Hey @henrikuper 

Can you please also share the `Interaction` model? Right now in the information that you have give, it doesn't contain this model so I can't try this reproduction locally. Thanks!

**Comment by henrikuper** [comment]
Hi @pantharshit00 I changed the interaction to test. It was just a typo. There is no interaction model.

**Comment by pantharshit00** [comment]
Hey I am unable to reproduce this in `2.4.1`.

Can you please try again with that version once?
![image](https://user-images.githubusercontent.com/22195362/90249598-5d2dea80-de58-11ea-8eb3-70288217fa6c.png)

**Comment by henrikuper** [comment]
Hi I tested it with 2.4.1 The error still occurs on my machine. 
`  Unique constraint failed on the fields: (`testId`)
    at PrismaClientFetcher.request (.../node_modules/@prisma/client/src/runtime/getPrismaClient.ts:1219:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Promise.all (index 1)
`

**Comment by pantharshit00** [comment]
@henrikuper Can you please try again with `2.5.1`?

**Comment by uncvrd** [comment]
For what it's worth, I ran in to this error when my sequence primary key id's were out of sync with the data in the table. I ran the following to see if each primary key id was correct, if not, I updated it...not sure if this'll fix your problem but here's what I did:

```
-- Login to psql and run the following

-- What is the result?
SELECT MAX(id) FROM "Token";

-- Then run...
-- This should be higher than the last result.
SELECT nextval('"Token_id_seq"');

-- If it's not higher... run this set the sequence last to your highest id. 
-- (wise to run a quick pg_dump first...)

BEGIN;
-- protect against concurrent inserts while you update the counter
LOCK TABLE "Token" IN EXCLUSIVE MODE;
-- Update the sequence
SELECT setval('"Token_id_seq"', COALESCE((SELECT MAX(id)+1 FROM "Token"), 1), false);
COMMIT;
```

**Comment by meotimdihia** [comment]
I have this problem: 
```
    for (const peopleId of peopleIds) {
      try {
        await prisma.md_comic_md_peoples.upsert({
          create: {
            md_comics: {
              connect: {
                id: id
              }
            },
            md_people: {
              connect: {
                id: peopleId
              }
            },
            type: type
          },
          update: {
            md_comics: {
              connect: { id: id }
            },
            md_people: {
              connect: {
                id: peopleId
              }
            },
            type: type
          },
          where: {
            A_B_type: {
              A: id,
              B: peopleId,
              type: type
            }
          }
        })
      } catch (e) {
        debugger
      }
    }
```

```
model md_comics {
  id                                             Int                   @id @default(autoincrement())
  mdid                                           Int?                  @unique
  title                                          String
  image                                          String?
  view_count                                     Int?
  follow_count                                   Int?
  rating                                         Float?
  bayesian_rating                                Float?
  rating_count                                   Int?
  chapter_count                                  Int?
  // This type is currently not supported.
  // created_at                                  timestamptz
  // This type is currently not supported.
  // updated_at                                  timestamptz
  // This type is currently not supported.
  // deleted_at                                  timestamptz?
  version                                        Int                   @default(0)
  alt_names                                      Json?
  genres                                         Json?
  authors                                        Json?
  artists                                        Json?
  desc                                           String?
  country                                        String?
  status                                         Int?
  hentai                                         Int?
  links                                          Json?
  slug                                           String?               @unique
  crawl_status                                   String?
  image_uploaded                                 String?
  image_gp_cdn                                   String?
  bing                                           Boolean               @default(false)
  last_volume                                    Float?
  last_chapter                                   Float?
  md_chapters                                    md_chapters[]
  relate_from                                    md_comic_md_comics[]  @relation("md_comic_md_comics_ATomd_comics")
  relate_to                                      md_comic_md_comics[]  @relation("md_comic_md_comics_BTomd_comics")
  md_genres                                      md_comic_md_genres[]
  md_people                                      md_comic_md_peoples[]
  md_covers                                      md_covers[]
  md_relateds_md_comicsTomd_relateds_md_comic_id md_relateds[]         @relation("md_comicsTomd_relateds_md_comic_id")
  md_relateds_md_comicsTomd_relateds_related_id  md_relateds[]         @relation("md_comicsTomd_relateds_related_id")
  md_titles                                      md_titles[]
  objs                                           objs[]

  @@index([title], name: "md_comics_gin_idx")
  @@index([title], name: "md_comics_title_idx")
  @@index([view_count], name: "md_comics_view_count_idx")
}

model md_people {
  id        Int                   @id @default(autoincrement())
  name      String                @unique
  slug      String
  md_comics md_comic_md_peoples[]
}

```

**Comment by pantharshit00** [comment]
Hmm, that is quite interesting @uncvrd 

Do you have any idea on what caused this for your case in the first place?

**Comment by uncvrd** [comment]
@pantharshit00 hey - so if I remember correctly, I deleted all items in my database (all the tables but didn't delete the database itself) and ran a migration to add all the same tables back, so I guess when running an upsert, the sequence ids from the old table did not reset. Hope that helps!

**Comment by guillaumeLamanda** [comment]
Hi there!  

I also hit this issue. I can't create a model instance with `upsert` when `where` have no property.
I made a small reproduction issue on CodeSandbox. 
https://codesandbox.io/s/prisma-upsert-reproduction-2tq0t?file=/src/index.js:389-394

This reproduction uses the latest version of prisma (`2.13.0`)

Hope that helps!

**Comment by pantharshit00** [comment]
@guillaumeLamanda 

I checked your reproduction and the output is expected. Since `id` property is not defined in the object that you are passing, it is `undefined` in javascript terms. If you pass `undefined` as a value to the client, it will omit that field: https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined

**Comment by Dremora** [comment]
I'm having a similar issue, both with `upsert` and `connectOrCreate`. Prisma 2.15.0.
In both cases, the key I'm using in the `where` clause is a unique index composed of two columns. I'm manually populating both values in the `create` clause (so they are not autoincremented values generated by the database). Of course in the real world the values are dynamic, but we had cases where the same set of values gets sent into our system nearly simultaneously.

Here's an example schema, code and logs (table and field names have been redacted). Even though this is `upsert` only (it's simpler), `connectOrCreate` generates similar transactions and queries, and results in the same issue.

```prisma
model Item {
  id                      Int                   @id @default(autoincrement())
  uniqueIndexField1       String                @map("unique_index_field_1")
  uniqueIndexField2       String                @map("unique_index_field_2")

  @@unique([uniqueIndexField1, uniqueIndexField2], name: "unique_index_field_1_unique_index_field_2_unique")
  @@map("items")
}
```


```typescript
async function addItem(field1, field2) {
  await prisma.item.upsert({
    where: {
      unique_index_field_1_unique_index_field_2_unique: {
        uniqueIndexField1: field1,
        uniqueIndexField2: field2
      }
    },
    create: {
      uniqueIndexField1: field1,
      uniqueIndexField2: field2
    },
    update: {}
  });
}

addItem('foo', 'bar');
addItem('foo', 'bar');

```

```
prisma:query BEGIN
prisma:query BEGIN
prisma:query SELECT "public"."items"."id" FROM "public"."items" WHERE ("public"."items"."unique_index_field_1" = $1 AND "public"."items"."unique_index_field_2" = $2) OFFSET $3
prisma:query SELECT "public"."items"."id" FROM "public"."items" WHERE ("public"."items"."unique_index_field_1" = $1 AND "public"."items"."unique_index_field_2" = $2) OFFSET $3
prisma:query INSERT INTO "public"."items" ("unique_index_field_1","unique_index_field_2") VALUES ($1,$25) RETURNING "public"."items"."id"
prisma:query SELECT "public"."items"."id","public"."items"."unique_index_field_1", "public"."items"."unique_index_field_2", FROM "public"."items" WHERE "public"."items"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
prisma:query INSERT INTO "public"."items" ("unique_index_field_1","unique_index_field_2") VALUES ($1,$2) RETURNING "public"."items"."id"
prisma:query ROLLBACK

(node:57819) UnhandledPromiseRejectionWarning: Error: 
Invalid `prisma.item.upsert()` invocation:


  Unique constraint failed on the fields: (`unique_index_field_1`,`unique_index_field_2`)
    at PrismaClientFetcher.request (node_modules/@prisma/client/runtime/index.js:78455:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

If you look at the logs, it becomes obvious what's happening. Two transactions start nearly simultaneously, both query `items` for a particular set of values, both determine that a new item needs to be created. So far database operations have been able to run in parallel. Then transaction 1 tries to create a new `item` in the database, transaction 2 has to wait. Once transaction 1 successfully commits, transaction 2 tries to create the same item - and fails.

Interestingly enough, we had similar code before with Knex, using hand-written transactions, and this rare-to-reproduce bug was present there as well.

**Comment by pantharshit00** [comment]
@Dremora I can reproduce the above issue. We should fix that.

**Comment by Dremora** [comment]
I'm not an expert in SQL, but I've been looking into potential solutions.
One would be to use `INSERT INTO ... ON CONFLICT` statement, that is basically an upsert. It only exists in Postgres. `UPSERT` can be used in MySQL instead.

An example from https://www.postgresqltutorial.com/postgresql-upsert/:

```sql
INSERT INTO customers (name, email)
VALUES('Microsoft','hotline@microsoft.com') 
ON CONFLICT (name) 
DO 
   UPDATE SET email = EXCLUDED.email || ';' || customers.email;
```

Any unique index key can be used after the `ON CONFLICT` statement.

Alternatively, database table would have to be locked using `ACCESS EXCLUSIVE` for reads before doing the first `SELECT`. This would ensure that the `SELECT`/`INSERT` pair is atomic. However, locking the whole table is not great for performance.

**Comment by matthewmueller** [comment]
Hey folks, this is expected behavior. 

If you issue two upserts concurrently in the Prisma Client, under the hood we do a read, then a write. This is to support datasources that don't have native upsert support. 

Unfortunately, this means that your concurrent queries have a race condition and may look liek this: read-read-create-create, the second create fails.

A couple workarounds:

It looks like you're just trying to create many records at once anyway, so you can use our [createMany API](https://www.prisma.io/docs/concepts/components/prisma-client/crud/#create-multiple-records) with `skipDuplicates: true`. Under the hood this does an `ON CONFLICT DO NOTHING` in Postgres.

You can also use upsert with await:

```ts
await addItem('foo', 'bar');
await addItem('foo', 'bar');
```

Finally you can upvote the [upsertMany issue](https://github.com/prisma/prisma/issues/5066). Community interest is one of the primary drives of features 🙂

**Comment by henrikuper** [comment]
Thanks @matthewmueller for the three ways workarounds, which sound very good if you have one of the three cases.

Actually I have a 4. alternative which causes this error, which can not be solved with one of these three solutions.

-> api endpoint is called twice at the exact same time with the same unique key. 

This happens nearly never in reality but theoretically it can happen. Both calls trigger the upsert function with the same unique key. Now the race condition read read create create happens and an error is thrown.

I solved the issue with a dirty hack by 
1. catching the error. 
2. then the catch function calls the upsert function again. 

This doesn't completely solve the issue, but reduces the probability of occurrence enough that there are no problems in the real application.

**Comment by matthewmueller** [comment]
Thanks @henrikuper, that's a great point about upsert across requests being racy. 

Highlighting @Dremora SQL to show that if two requests come in at the same time trying to update the same keys, you can end up with a unique constraint violation

```ts
prisma:query BEGIN
prisma:query BEGIN
prisma:query SELECT "public"."items"."id" FROM "public"."items" WHERE ("public"."items"."unique_index_field_1" = $1 AND "public"."items"."unique_index_field_2" = $2) OFFSET $3
prisma:query SELECT "public"."items"."id" FROM "public"."items" WHERE ("public"."items"."unique_index_field_1" = $1 AND "public"."items"."unique_index_field_2" = $2) OFFSET $3
prisma:query INSERT INTO "public"."items" ("unique_index_field_1","unique_index_field_2") VALUES ($1,$25) RETURNING "public"."items"."id"
prisma:query SELECT "public"."items"."id","public"."items"."unique_index_field_1", "public"."items"."unique_index_field_2", FROM "public"."items" WHERE "public"."items"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
prisma:query INSERT INTO "public"."items" ("unique_index_field_1","unique_index_field_2") VALUES ($1,$2) RETURNING "public"."items"."id"
prisma:query ROLLBACK
```

I've updated the title and re-opened this issue. Thanks for sharing this key context @henrikuper!

**Comment by windowsdeveloperwannabe** [comment]
Should Prisma's `upsert` not being a real upsert be highlighted by the documentation? Or alternatively the method renamed to `create_or_update` like in ActiveRecord. 🤔 It is quite the blocker when you start getting unique constraint errors from simple upserts.

Prisma's data guide pages even describe the real upsert so Google results can get quite confusing.

**Comment by anglinb** [comment]
👋  Any update here? I also ended up hacking together a try-catch solution like @henrikuper but it's very gross.

**Comment by elie222** [comment]
Pretty annoying issue. I run into it with `connectOrCreate`.
The issue is that the parent record isn't created either. I could just ignore the error for the related record as it's already been added. But it causes the entire operation to fail which I can't have :(

**Comment by janpio** [comment]
DataGuide != Docs. The DataGuide describes raw PostgreSQL, not Prisma Queries. You can find the Prisma docs at https://www.prisma.io/docs/.

**Comment by bartoszhernas** [comment]
I mentioned data guide because directly underneath the information about how to use/code ON CONFLICT in Postgres clause there is a mention

> Note: If you are connecting to your database with Prisma client, you can perform upsert operations using the dedicated upsert operation.

which links to prisms docs. It is highly misleading but I still think prisma in Postgres should just use ON CONFLICT clause. Just as data guide suggest.

**Comment by mahdavipanah** [comment]
> Hey folks, this is expected behavior.
> 
> If you issue two upserts concurrently in the Prisma Client, under the hood we do a read, then a write. This is to support data sources that don't have native upsert support.
> 
> Unfortunately, this means that your concurrent queries have a race condition and may look like this: read-read-create-create, and in this case, the second create fails the constraint.
> 
> **A couple workarounds:**
> 
> 1. It looks like you're just trying to create many records at once anyway, so you can use our [createMany API](https://www.prisma.io/docs/concepts/components/prisma-client/crud/#create-multiple-records) with `skipDuplicates: true`. Under the hood this does an `ON CONFLICT DO NOTHING` in Postgres.
> 2. You can also use upsert with await:
> 
> ```ts
> await addItem('foo', 'bar');
> await addItem('foo', 'bar');
> ```
> 
> 3. Finally you can upvote the [upsertMany issue](https://github.com/prisma/prisma/issues/5066). Community interest is one of the primary drives of features 🙂

I think this behavior must be documented. Because the intuitive inference is that this operation is atomic.  I am wondering if wrapping the `upsert` operation in `prisma.$transaction` will solve the issue:
```javascript
prisma.$transaction([
    prisma.myModel.upsert({...})
])
```

**Comment by alonbilu** [comment]
Experiencing the same problem with a compound key... I too believe something should be done about it. Having to use the directy try/catch hack as the only solution is very lame :(

**Comment by bartoszhernas** [comment]
Who though that doing completely opposite what the DataGuide recommends is a good choice?

I understand rationale for supporting other mechanism that do not support native UPSERT, but then just make it clear in Docs, and underneath use whatever is best suited for each type of DB. You are using ForeignKeys in Postgres, even though MongoDB doesn't have it right?

> Prisma's Data Guide
Learn how databases work, how to choose the right one, and how to use databases with your applications to their full potential.

Ok, so let me learn, but then do NOT allow me to use the full potential of the DB, at the same time misleading me of what upsert does (dataguide says to use prisma's upsert as example of how to use postgre's UPSERT, which is a lie). What gives? You are teaching people to do X, and at the same time you do not do the X. You even make it hard/impossible to do the correct thing while using Prisma :(

**Comment by Noor0** [comment]
I'm facing the same issue while trying to call `await createMany(...)` after a call to `await deleteMany(...)` in a `prisma.$transaction` and I'm not calling `upsert` at all.

**Comment by janpio** [comment]
@Noor0 That case is probably better represented in https://github.com/prisma/prisma/issues/9678
@elie222 @Dremora Are you aware of any issues that spell out this problem with examples of `connectOrCreate` usage?

**Comment by janpio** [comment]
Ping @elie222 @Dremora Any `connectOrCreate` examples would be super helpful for us. Thanks.

**Comment by dtinth** [comment]
I got bitten by this too. I submitted a documentation change to add the following remark to `upsert`’s API docs:

> Prisma performs the upsert first executing a `SELECT` query to check if the record exists, followed by an `INSERT` or `UPDATE` query depending on the result of the `SELECT` query. Since it does not use the `ON CONFLICT` or the `ON DUPLICATE KEY` clause, [two simultaneous upserts with the same `UNIQUE` key can result in a “Unique constraint failed” error](https://github.com/prisma/prisma/issues/3242). It is your application’s responsibility to handle the [P2002](error-reference#p2002) error and retry the upsert.

PR:

- https://github.com/prisma/docs/pull/3910

**Comment by dickfickling** [comment]
🙌

**Comment by garrensmith** [comment]
Prisma Upserts are very powerful and support very nested upserts. This can lead to a Unique constraint error being thrown. We have updated our [docs](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#unique-key-constraint-errors-on-upserts) with why this happens and what to do in that situation. 
We have also added support for `INSERT .. ON CONFLICT .. UPDATE` see the [docs](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#database-upserts) on when it will be used.

**Comment by t0ggah** [comment]
@garrensmith: Great that it's documented better. We have the case([which I see the example also have](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#database-upsert-examples)), that we have unique constraint with the combination of two fields. 
Would a where clause combining these two use native or prisma upsert? E.g.
```prisma
prisma.User.upsert({
  where: {
   userName_profileViews: {
      userName: 'Alice',
      profileViews: 1,
    }
  },
  create: {
   ...
  },
  update: {
    ...
  },
})

**Comment by garrensmith** [comment]
@t0ggah yes if your create has the two fields defined defined in the where clause and they are the same

```
prisma.User.upsert({
  where: {
   userName_profileViews: {
      userName: 'Alice',
      profileViews: 1,
    }
  },
  create: {
    userName: 'Alice',
     profileViews: 1,
  },
  update: {
    ...
  },
})
```

**Comment by rothfels** [comment]
@garrensmith thanks for adding real upserts -- super helpful. Any reason you're not supporting MySQL `ON DUPLICATE KEY`, or plans to in the future? We (MySQL users) want true upserts too :)

**Comment by rothfels** [comment]
@janpio are database upserts for MySQL on the product roadmap? Referring to those using `ON DUPLICATE KEY` (not the currently supported prisma upsert functionality). Thanks.

**Comment by janpio** [comment]
I fear that dropped between the other work. Can you open a feature request for this? Generally I see no reason why we should not also do similar changes/optimizations (if I am allowed to call them that) for MySQL 👍

**Comment by felipap** [comment]
@garrensmith I believe the following query satisfies the requirements to use a native upsert:

```ts
const schema = await db.destinationSchema.upsert({
  where: {
    destinationId: dest.id,
  },
  create: {
    destinationId: dest.id,
    envId: dest.envId,
    engine: dest.engine,
    projectId: dest.projectId,
  },
  update: {},
})
```

But logs show a multi-stage upsert taking place:
```
prisma:query BEGIN
prisma:query SELECT "public"."DestinationSchema"."id" FROM "public"."DestinationSchema" WHERE ("public"."DestinationSchema"."id" = $1 AND 1=1) OFFSET $2
prisma:query INSERT INTO "public"."DestinationSchema" ("id","createdAt","updatedAt","destinationId","engine","envId","projectId") VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING "public"."DestinationSchema"."id"
prisma:query SELECT "public"."DestinationSchema"."id", "public"."DestinationSchema"."createdAt", "public"."DestinationSchema"."updatedAt", "public"."DestinationSchema"."destinationId", "public"."DestinationSchema"."value", "public"."DestinationSchema"."initializedAt", "public"."DestinationSchema"."engine", "public"."DestinationSchema"."envId", "public"."DestinationSchema"."projectId" FROM "public"."DestinationSchema" WHERE "public"."DestinationSchema"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
```

Am I missing something?

- [x] You use Prisma version 4.6.0 or later => Yes, 4.16.
- [x] Your application uses a CockroachDB, PostgreSQL, or SQLite data source => Yes, postgres.
- [x] The query does not include a selection that uses a nested read => true
- [x] The query modifies only one model => true
- [x] There is only one unique field in the upsert's where option => true, destinationId
- [x] The unique field in the where option and the unique field in the create option have the same value => true

**Comment by janpio** [comment]
Can you please open a new issue @felipap? We'll be happy to reproduce and look into that. Thanks!

**Comment by rothfels** [comment]
@felipap I discovered when the `update` is empty the native upsert is not used, even if the rest of the rules are followed. This is an odd choice (or just a bug) but you can work around it by adding `destinationId: dest.id,` to the update.

**Comment by bartoszhernas** [comment]
Was this issue closed because it was fixed, if so which version? As @rothfels mentioned, in case the `update` is empty the native upsert is not used which is not what the docs require for native upsert to be used.

**Comment by janpio** [comment]
Then please open a new issue and describe the full situation. Thanks.

**Comment by ideosyncretic** [comment]
New issue here, for anyone still following this: https://github.com/prisma/prisma/issues/21853
